src/lib.rs

```diff
@@ -342,15 +343,43 @@ where
             continue;
         }
         visited.insert(start);
-        let mut dfs = Dfs::new(graph, start);
-        let mut parent: NodeIndex = start;
-        while let Some(nx) = dfs.next(graph) {
-            if visited.contains(&nx) {
-                continue;
+        let mut children: Vec<NodeIndex> = graph.neighbors(start).collect();
+        children.reverse();
+        let mut stack: Vec<(NodeIndex, Vec<NodeIndex>)> =
+            vec![(start, children)];
+        // Used to track the last position in children vec across iterations
+        let mut index_map: HashMap<NodeIndex, usize> = HashMap::new();
+        index_map.insert(start, 0);
+        while !stack.is_empty() {
+            let temp_parent = stack.last().unwrap();
+            let parent = temp_parent.0;
+            let children = temp_parent.1.clone();
+            let count = *index_map.get(&parent).unwrap();
+            let mut found = false;
+            let mut index = 0;
+            for child in &children {
+                // Skip children for a parent already inspected.
+                if index < count {
+                    index += 1;
+                    continue;
+                }
```

Comment:
Could you set `index` to `count` instead?

Reply:
I could, the tricky bit is because this is iterating over the list. To do that I'd probably have to change the for loop to be a while loop an manually access each element. Something like:

```rust
while index < children.len() {
    child = children[index]
    if index < count {
        index = count;
        continue;
...
}
```

Reply:
Oh, as you mentioned in slack using a slice here is a better path than switching to a while loop. I tested this locally and it works fine. I've pushed this up in: https://github.com/Qiskit/retworkx/pull/146/commits/1f4caafbf65b6ba3aa5db7e4152d4f02d91d7d87 which removes this if statement.